contracts/Uniswap/UniswapAnchoredView.sol

```diff
@@ -0,0 +1,241 @@
+     * @param cToken The cToken address for price retrieval
+     * @return The price for the given cToken address
+     */
+    function getUnderlyingPrice(address cToken) public view returns (uint) {
+        TokenConfig memory config = getTokenConfigByCToken(cToken);
+        return priceInternal(config);
+    }
+
+    /**
+     * @notice Post open oracle reporter prices, and recalculate stored price by comparing to anchor
+     * @dev We let anyone pay to post anything, but only prices from configured reporter will be stored in the view.
+     * @param messages The messages to post to the oracle
+     * @param signatures The signatures for the corresponding messages
+     * @param symbols The symbols to compare to anchor for authoritative reading
+     */
+    function postPrices(bytes[] calldata messages, bytes[] calldata signatures, string[] calldata symbols) external {
+        require(messages.length == signatures.length, "messages and signatures must be 1:1");
+
+        // Save the prices
+        for (uint i = 0; i < messages.length; i++) {
+            priceData.put(messages[i], signatures[i]);
+        }
+
+        uint ethPrice = fetchAnchorPrice(getTokenConfigBySymbol("ETH"), 1e6);
+
+        // Try to update the view storage
+        for (uint i = 0; i < symbols.length; i++) {
+            TokenConfig memory config = getTokenConfigBySymbol(symbols[i]);
+            string memory symbol = symbols[i];
+            bytes32 symbolHash = keccak256(abi.encodePacked(symbol));
+            if (source(messages[i], signatures[i]) != reporter) continue;
+
+            uint reporterPrice = priceData.getPrice(reporter, symbol);
+            uint anchorPrice = fetchAnchorPrice(config, ethPrice);
+
+            uint anchorRatio = mul(anchorPrice, 100e16) / reporterPrice;
+            bool withinAnchor = anchorRatio <= upperBoundAnchorRatio && anchorRatio >= lowerBoundAnchorRatio;
+
+            if (reporterInvalidated == true) {
+                if (prices[symbolHash] == anchorPrice) {
+                    prices[symbolHash] = anchorPrice;
+                }
+            } else if (withinAnchor) {
+                if (prices[symbolHash] != reporterPrice) {
+                    prices[symbolHash] = reporterPrice;
+                    emit PriceUpdated(symbol, reporterPrice);
+                }
+            } else {
+                emit PriceGuarded(symbol, reporterPrice, anchorPrice);
+            }
+        }
+    }
+
+    /**
+     * @dev Fetches the token/eth price accumulator for a config, with same decimals as the token.
+     */
+    function currentCumulativePrice(TokenConfig memory config) internal view returns (uint) {
+        (uint price0Cumulative, uint price1Cumulative,) = UniswapV2OracleLibrary.currentCumulativePrices(config.uniswapMarket);
+        if (config.isUniswapReversed) {
+            return mul(price0Cumulative, 1e18 / config.baseUnit);
+        } else {
+            return mul(price1Cumulative, config.baseUnit / 1e18);
```

Comment:
probably need to do the mul first , since this could truncate to 0 eg 1e6 / 1e18, right?

Reply:
probably, good point